### PR TITLE
dependencies: bump LuaJIT to 2.1 branch HEAD @ e9af1ab.

### PR DESCRIPTION
**Pull request by htuch**
LuaJIT 2.1.0-beta3 has the following CVEs, which don't appear
super critical for correctly functioning Lua code but prudence dictates
we should bump anyway:

- CVE-2020-15890: LuaJit through 2.1.0-beta3 has an out-of-bounds read
  because __gc handler frame traversal is mishandled.

- CVE-2020-24372: LuaJIT through 2.1.0-beta3 has an out-of-bounds read
  in lj_err_run in lj_err.c.

There is no release version beyond 2.1.0-beta3, so using HEAD of 2.1
branch.

Risk level: Medium (if using LuaJIT).
Testing: bazel test //test/... Some unit tests required fixups due to
  changes in Lua header map table ordering.

Signed-off-by: Harvey Tuch <htuch@google.com>
Co-authored-by: Michael Payne <michael@sooper.org>
Co-authored-by: William A Rowe Jr <wrowe@vmware.com>

### CODE REVIEW

**Comment by mattklein123** [comment]
Check CI and I can take a look?

/wait

**Comment by wrowe** [comment]
I'm on it, my boxes were all spent fighting with the last of the clang-cl issues (much simpler if I was on master and not a fork of 13133, which has been lingering too long.) The patch is ready, I'll be at the office Sunday to hand off the patch of our localization plus their improvements to the DASM code logic for x64.

In other events, we discussed the issue of the unmaintained moonjit fork and this patch. It's clear, we must move to either CMake or bazel and push that upstream to luajit, decommission the moonjit support, and find someone (perhaps @iii-i ?) to guide s390 porting patches back to LuaJiT upstream. I will open the ticket for envoy, draft a portable substitute for this horrid patched .bat makefile, and help shepherd whatever we can accomplish with the LuaJIT maintainers.

**Comment by htuch** [comment]
@wrowe LMK when you have the Windows patch sorted and I'll apply. Seems there is an orthogonal gcc failure on the wrappers_test, so I'll look into that meanwhile.

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -237,13 +237,15 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
-        version = "2.1.0-beta3",
-        sha256 = "409f7fe570d3c16558e594421c47bdd130238323c9d6fd6c83dedd2aaeb082a8",
+        # The last release version, 2.1.0-beta3 has a number of CVEs filed
+        # against it. These may not impact correct non-malicious Lua code, but for prudence we bump.
+        version = "e9af1abec542e6f9851ff2368e7f196b6382a44c",
+        sha256 = "186cd55c0e321aa2ec3c0f9220e31cc17479ec6222db515ec8b902f9275a9616",
         strip_prefix = "LuaJIT-{version}",
-        urls = ["https://github.com/LuaJIT/LuaJIT/archive/v{version}.tar.gz"],
+        urls = ["https://github.com/LuaJIT/LuaJIT/archive/{version}.tar.gz"],
+        last_updated = "2020-10-09",
```
Comment (moderation) [inline bazel/repository_locations.bzl:246]: This commit is 2020-09-29 and `last_updated` should reflect that. Consider bumping to latest commit `1d8b747c161db457e032a023ebbff511f5de5ec2` `2020-10-12`.

File: bazel/foreign_cc/luajit.patch [hunk 2]
```diff
@@ -33,96 +33,107 @@ index f56465d..5d91fa7 100644
...
+ @if errorlevel 1 goto :BAD
+ %LJLINK% /DLL /out:%LJDLLNAME% ljamalg.obj lj_vm.obj
+ @if errorlevel 1 goto :BAD
 diff --git a/build.py b/build.py
 new file mode 100755
-index 0000000..9c71271
+index 0000000..3eb74ff
 --- /dev/null
 +++ b/build.py
 @@ -0,0 +1,56 @@
```
Comment (wrowe) [inline bazel/foreign_cc/luajit.patch:139]: @htuch the patch line 182 below;
```
os.system('msvcbuild.bat gc64 ' + os.getenv('WINDOWS_DBG_BUILD', '') + ' static')
```
must be changed to omit the gc64 flag (which became a nogc64 toggle on their 2.1-branch). E.g.
```
os.system('msvcbuild.bat ' + os.getenv('WINDOWS_DBG_BUILD', '') + ' static')
```
(same reasoning as dropping the `@@ -110,8 +110,8 @@` patch above.)

**Review by wrowe — CHANGES_REQUESTED** [review]
Still locally encountering various build breakage even after feature define wasm=disabled, merge to master may or may not help, but the patch suggestions appear correct, thanks to @sunjayBhatia for diagnostics

**Comment by mattklein123** [comment]
Merge main to fix the checks?

/wait

**Review by wrowe — APPROVED** [review]
LGTM

**Pull request merged**